### PR TITLE
Issue #11357: Report correct violation in lambda return

### LINKED ISSUE #11357: UnnecessaryParentheses reports incorrect violation message for return value in lambdas
I have read check documentation: https://checkstyle.org/config_coding.html#UnnecessaryParentheses
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```shell
$ javac ParenthesesTest.java
# ParenthesesTest.class created

$ type config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name = "Checker">
        <module name="TreeWalker">
                <module name="UnnecessaryParentheses"/>
        </module>
</module>

$ type ParenthesesTest.java
import java.util.function.Function;

public class ParenthesesTest {
    public void returnValueExample() {
        Function<Integer, Integer> addFive = x -> {
            return (x + 5); // violation, should be reported for parentheses around return value
        };
    }
}

$ java "-Duser.language=en -Duser.country=US" -jar checkstyle-10.0-all.jar -c config.xml ParenthesesTest.java
Starting audit...
[ERROR] C:\Users\Faraaz\CheckstyleTest\ParenthesesTest.java:6:20: Unnecessary parentheses around assignment right-hand side. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 1 errors.
```
---

When a lambda expression is assigned to a variable and has a return value surrounded in parentheses, the violation reported is ' Unnecessary parentheses around assignment right-hand side.'.

**Expected:**
The correct violation message should be 'Unnecessary parentheses around return value.'.
ParenthesesTest.java:6:20 should be reported for unnecessary parentheses around return value.

---

Could it be due to [this](https://github.com/checkstyle/checkstyle/blob/4981aad0c6a4270f618fcab1b50a00c0c22a33f0/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java#L585), wherein the assignment depth is being checked **before** checking whether the parent token is a 'return'?

### PR DESCRIPTION
**Pull request by OttoKaaij**
Fixes #11357. Had to fix some tests, also in the SwitchExpression test

Diff Regression config: https://gist.githubusercontent.com/OttoKaaij/363ce8a2a1007b4a3856cda24bb4e1c6/raw/4e6e0ec1c7b56bc728034e36ec498527c7aa2f8d/my_checks.xml

Regression report: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/cdec517_2022140934/reports/diff/index.html

### CODE REVIEW

**Comment by OttoKaaij** [comment]
Working on fixing `mvm clean verify`

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckSwitchExpression.java [hunk 1]
```diff
@@ -46,16 +46,16 @@ MathOp2 tooManyParens2(int k) {
         MathOp2 e = (a, b) -> (a + b); // violation 'parentheses around assignment right-hand side'
       }
       case (2) -> { // violation 'Unnecessary parentheses around literal '2''
-        MathOp2 f = (int a, int b) -> (a + b); // violation 'paren.* around assignment right.*side'
+        MathOp2 f = (int a, int b) -> (a + b); // violation 'paren.* around assignment right.* side'
```
Comment (nick-mancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckSwitchExpression.java:49]: Please revert change here, or make a separate `minor` commit for formatting of violation message; this is not related to check update.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java [hunk 2]
```diff
@@ -15,7 +15,7 @@
 public class InputUnnecessaryParenthesesLambdas {
     int foo(int y) {
         MathOperation case1 = (x) -> x + x; // violation 'Unnecessary paren.* around lambda value'
-        MathOperation case2 = (x) -> { return x + x; }; // violation 'paren.* around lambda value'
+        MathOperation case2 = (x) -> { return ((x + x)); }; // violation 'paren.* around lambda.*'
```
Comment (nick-mancuso) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java:18]: Please do not modify existing test cases, except to change violation message.

**Review by nick-mancuso — CHANGES_REQUESTED** [review]
Please generate check regression report, you can find details here: https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#report-generation

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java (hunk 2 above)
Comment (OttoKaaij) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java:18]: Fixed

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckSwitchExpression.java (hunk 1 above)
Comment (OttoKaaij) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckSwitchExpression.java]: fixed

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java [hunk 3]
```diff
@@ -15,7 +15,7 @@
 public class InputUnnecessaryParenthesesLambdas {
     int foo(int y) {
         MathOperation case1 = (x) -> x + x; // violation 'Unnecessary paren.* around lambda value'
-        MathOperation case2 = (x) -> { return x + x; }; // violation 'paren.* around lambda value'
+        MathOperation case2 = (x) -> { return x + x; };
```
Comment (nick-mancuso) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java:18]: Why is this not a violation anymore?

**Review by nick-mancuso — CHANGES_REQUESTED** [review]

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java (hunk 3 above)
Comment (OttoKaaij) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java]: It is still a violation, my mistake. Reverted

**Comment by OttoKaaij** [comment]
@nick-mancuso Before I generate the regression report, could you check if the config is correct like this?

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckSwitchExpression.java (hunk 1 above)
Comment (nick-mancuso) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckSwitchExpression.java]: Resolved.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java (hunk 2 above)
Comment (nick-mancuso) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java]: Resolved.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java (hunk 3 above)
Comment (nick-mancuso) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java]: Resolved.

**Comment by nick-mancuso** [comment]
> @nick-mancuso Before I generate the regression report, could you check if the config is correct like this?

Config looks good, you don't have to specify the projects file, we have default for github action.

**Comment by OttoKaaij** [comment]
GitHub, generate report

**Comment by nick-mancuso** [comment]
Please extend your new input file to include cases like:

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/cdec517_2022140934/reports/diff/apache-ant/xref/home/runner/work/checkstyle/checkstyle/contribution/checkstyle-tester/repositories/apache-ant/src/tests/junit/org/apache/tools/ant/taskdefs/XmlPropertyTest.java.html#L360

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/cdec517_2022140934/reports/diff/checkstyle/xref/home/runner/work/checkstyle/checkstyle/contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationValidClassDefIndent.java.html#L144

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/cdec517_2022140934/reports/diff/hibernate-orm/xref/home/runner/work/checkstyle/checkstyle/contribution/checkstyle-tester/repositories/hibernate-orm/hibernate-core/src/test/java/org/hibernate/test/bytecode/enhancement/customer/CustomerEnhancerTest.java.html#L100

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/cdec517_2022140934/reports/diff/spring-framework/xref/home/runner/work/checkstyle/checkstyle/contribution/checkstyle-tester/repositories/spring-framework/spring-tx/src/main/java/org/springframework/transaction/interceptor/TransactionAttributeSourceAdvisor.java.html#L46

**Comment by OttoKaaij** [comment]
> Please extend your new input file to include cases like:

Should be done!

**Comment by nick-mancuso** [comment]
Ci failure is not related, but is consistent: https://checkstyle.semaphoreci.com/jobs/92d3c7fd-7382-4984-991a-4ee4eb4da260

`Execution default of goal org.antlr:antlr4-maven-plugin:4.9.3:antlr4 failed: A required class was missing while executing org.antlr:antlr4-maven-plugin:4.9.3:antlr4: org/stringtemplate/v4/STGroup`

I suppose this is a problem downloading some ANTLR dependency.

**Review by nick-mancuso — APPROVED** [review]

**Review by strkkk — APPROVED** [review]

**Review by pbludov — APPROVED** [review]

**Review by romani — APPROVED** [review]

**Pull request merged**